sdk/go/common/resource/testing/rapid.go

```diff
@@ -412,3 +460,235 @@ func propertyValueGenerator(ctx *StackContext, maxDepth int) *rapid.Generator {
+
+		// Draw inputs and calculate property dependencies from any nested output values therein.
+		inputs := propertyMapGenerator(ctx, maxDepth).Draw(t, "inputs").(resource.PropertyMap)
+		propertyDependencies := map[resource.PropertyKey][]resource.URN{}
+		for k, v := range inputs {
+			propertyDependencySet := map[resource.URN]bool{}
+			getValueDependencies(propertyDependencySet, v)
+			unionDeps(dependencySet, propertyDependencySet)
+			if len(propertyDependencySet) > 0 {
+				propertyDependencies[k] = sortedDeps(propertyDependencySet)
+			}
+		}
+
+		// Draw outputs. If this is a custom resource, remove nested deps and sample secret outputs.
+		outputs := propertyMapGenerator(ctx, maxDepth).Draw(t, "outputs").(resource.PropertyMap)
+		var additionalSecretOutputs []resource.PropertyKey
+		if custom {
+			removeOutputDeps(resource.NewObjectProperty(outputs))
+
+			additionalSecretOutputs = make([]resource.PropertyKey, 0, len(outputs))
+			for k := range outputs {
+				additionalSecretOutputs = append(additionalSecretOutputs, k)
+			}
+
+			count := rapid.IntRange(0, len(additionalSecretOutputs)).Draw(t, "additionalSecretOutputs").(int)
+			additionalSecretOutputs = additionalSecretOutputs[:count]
+
+			rand.Shuffle(len(additionalSecretOutputs), func(i, j int) {
+				additionalSecretOutputs[i], additionalSecretOutputs[j] = additionalSecretOutputs[j], additionalSecretOutputs[i]
+			})
+		}
+
+		// If this is a custom resource, generate init errors.
+		var initErrors []string
+		if custom && !isProvider {
+			initErrors = rapid.SliceOfN(rapid.String(), 0, 4).Draw(t, "init errors").([]string)
+		}
+
+		// If this is a non-provider custom resource, generate a provider reference.
+		var provider string
+		if custom && !isProvider {
+			provider = string(providerResource.URN) + "::" + string(providerResource.ID)
+		}
+
+		// If this is a custom resource, generate custom timeouts.
+		var timeouts resource.CustomTimeouts
+		if custom && rapid.Bool().Draw(t, "custom timeouts").(bool) {
+			timeouts.Create = rapid.OneOf(rapid.Just(0.0), rapid.Float64Max(100000)).Draw(t, "create timeout").(float64)
+			timeouts.Update = rapid.OneOf(rapid.Just(0.0), rapid.Float64Max(100000)).Draw(t, "update timeout").(float64)
+			timeouts.Delete = rapid.OneOf(rapid.Just(0.0), rapid.Float64Max(100000)).Draw(t, "delete timeout").(float64)
+		}
+
+		return &resource.State{
+			Type:                    urn.Type(),
+			URN:                     urn,
+			Custom:                  custom,
+			Delete:                  del,
+			ID:                      id,
+			Inputs:                  inputs,
+			Outputs:                 outputs,
+			Parent:                  parent,
+			Protect:                 protect,
+			External:                external,
+			Dependencies:            sortedDeps(dependencySet),
+			InitErrors:              initErrors,
+			Provider:                provider,
+			PropertyDependencies:    propertyDependencies,
```

Comment:
Curious the code above makes it look like PropertyDependencies is a computed field fully determined by Inputs. We chose to include it in the model for redundancy?

Reply:
> Curious the code above makes it look like PropertyDependencies is a computed field fully determined by Inputs. We chose to include it in the model for redundancy?

Until @justinvp's recent work, we did not have a first-class representation of `Output` values in the `resource.PropertyValue` system, so all dependency information was lifted to the top level during marshaling. We can probably have the SDKs stop producing this field if the resource monitor they're connected to supports `Output` values. We could probably also move the responsibility for consuming these values into the resource monitor from the engine (e.g. the resource monitor could wrap any top-level property in `Inputs` that has a non-empty set of dependencies in `PropertyDependencies` with an `Output` value that has the same set of dependencies).

Reply:
Ah that totally makes sense. I wouldn't change this just yet until the new changes "settle in" and become the new normal so to speak.